doc/sphinx/proof-engine/tactics.rst

```diff
@@ -1522,144 +1471,107 @@ Managing the local context
-   of name :n:`H` asserting :g:`U` to the current goal and opens a new subgoal
-   :g:`U` [2]_. The subgoal :g:`U` comes first in the list of subgoals remaining to
-   prove.
-
-   .. exn:: Not a proposition or a type.
-
-      Arises when the argument :token:`type` is neither of type :g:`Prop`,
-      :g:`Set` nor :g:`Type`.
-
-   .. tacv:: assert @type
+.. tacn:: assert ( @ident : @term ) {? by @ltac_expr3 }
+          assert @one_term {? @as_ipat } {? by @ltac_expr3 }
+   :name: assert; _
 
-      This behaves as :n:`assert (@ident : @type)` but :n:`@ident` is
-      generated by Coq.
+   The first form adds a new hypothesis named :n:`@ident` containing :n:`@term` and
+   creates a new subgoal, which becomes the first in the list of subgoals
+   remaining to prove.  (This corresponds to the cut rule of sequent calculus.)
 
-   .. tacv:: assert @type by @tactic
+   The second form adds a new hypothesis with a name chosen by Coq
+   or specified by :n:`@as_ipat`.  If :n:`@ltac_expr3` is specified, it is
+   applied to the new subgoal.
 
-      This tactic behaves like :tacn:`assert` but applies tactic to solve the
-      subgoals generated by assert.
+   :n:`@as_ipat`
+     Gives the name for the hypothesis or specifies how to split it into parts.
+     See :ref:`intropatterns`.
 
-      .. exn:: Proof is not complete.
-         :name: Proof is not complete. (assert)
-         :undocumented:
+   :n:`by @ltac_expr3`
+     Applies :n:`@ltac_expr3` to solve the subgoal generated by :n:`assert`.
 
-   .. tacv:: assert @type as @simple_intropattern
+     If not specified, the tactic applies `exact @term` by default where
+     :token:`type` is the type of :token:`term` and also equivalent to
+     :tacn:`pose proof`.  If the head of term is :token:`ident`, the tactic
+     is equivalent to :tacn:`specialize`.
 
-      If :n:`simple_intropattern` is an intro pattern (see :ref:`intropatterns`),
-      the hypothesis is named after this introduction pattern (in particular, if
-      :n:`simple_intropattern` is :n:`@ident`, the tactic behaves like
-      :n:`assert (@ident : @type)`). If :n:`simple_intropattern` is an action
-      introduction pattern, the tactic behaves like :n:`assert @type` followed by
-      the action done by this introduction pattern.
-
-   .. tacv:: assert @type as @simple_intropattern by @tactic
-
-      This combines the two previous variants of :tacn:`assert`.
+   .. exn:: Not a proposition or a type.
 
-   .. tacv:: assert (@ident := @term)
+      Arises when the argument :token:`type` is neither of type :g:`Prop`,
+      :g:`Set` nor :g:`Type`.
 
-      This behaves as :n:`assert (@ident : @type) by exact @term` where
-      :token:`type` is the type of :token:`term`. This is equivalent to using
-      :tacn:`pose proof`. If the head of term is :token:`ident`, the tactic
-      behaves as :tacn:`specialize`.
+   .. exn:: Proof is not complete.
+      :name: Proof is not complete. (assert)
+      :undocumented:
```

Comment:
This happens when somebody does
```
assert (x : P) by tactic.
```
the idea is that it doesn't produce a goal, so if tactic doesn't solve `P` then it gives this error.